File: franchise_erp/franchise_erp/report/sales_vs_stock/sales_vs_stock.js
// Copyright (c) 2026, Franchise Erp and contributors
// For license information, please see license.txt



frappe.query_reports["Sales vs Stock"] = {
    filters: [
        {
            fieldname: "from_date",
            label: "From Date",
            fieldtype: "Date",
            default: frappe.datetime.month_start()
        },
        {
            fieldname: "to_date",
            label: "To Date",
            fieldtype: "Date",
            default: frappe.datetime.month_end()
        },
        {
            fieldname: "month",
            label: "Month",
            fieldtype: "Select",
            options: [
                { "value": "", "label": "All" },
                { "value": "1",  "label": "Jan" },
                { "value": "2",  "label": "Feb" },
                { "value": "3",  "label": "Mar" },
                { "value": "4",  "label": "Apr" },
                { "value": "5",  "label": "May" },
                { "value": "6",  "label": "Jun" },
                { "value": "7",  "label": "Jul" },    // document.addEventListener(
                { "value": "8",  "label": "Aug" },
                { "value": "9",  "label": "Sep" },
                { "value": "10", "label": "Oct" },
                { "value": "11", "label": "Nov" },
                { "value": "12", "label": "Dec" }
            ],
            default: ""
        },
        {
            fieldname: "metric",
            label: "View By",
            fieldtype: "Select",
            options: ["qty", "amt"],  
            default: "qty"
        },
        {
            fieldname: "company",
            label: "Company",
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company")
        }
    ],

    // onload: function(report) {
    //     if (!report.page) return;
    //     if (!report.get_filter_value("month")) {
    //         report.set_filter_value("month", (new Date().getMonth() + 1).toString());
    //     }
    // },
onload: function(report) {

    if (report._dashboard_listener_added) {
        return;
    }

    report._dashboard_listener_added = true;

    document.addEventListener(
        "dashboardFilterChanged",
        function(e) {

            const f = e.detail;

            if (report.get_filter("from_date")) {
                report.set_filter_value("from_date", f.from);
            }

            if (report.get_filter("to_date")) {
                report.set_filter_value("to_date", f.to);
            }

            if (report.get_filter("view_type")) {
                report.set_filter_value("view_type", f.view);
            }

            if (report.get_filter("metric")) {
                report.set_filter_value("metric", f.view);
            }

            if (report.get_filter("company")) {
                report.set_filter_value("company", f.company || "");
            }

            report.refresh();
        }
    );
},

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "status" && data) {
            if (data.status === "Low Stock") {
                value = `<span style="color:red;font-weight:bold;">${value}</span>`;
            } else if (data.status === "In Stock") {
                value = `<span style="color:green;">${value}</span>`;
            }
        }

        return value;
    },

    get_chart_data: function(columns, result) {

        if (!result || result.length === 0) {
            return null;
        }

        let labels = [];
        let sales = [];
        let stock = [];

        result.forEach(row => {

            labels.push(
                row.custom_barcode_code ||
                row.item_name ||
                "Unknown"
            );

            sales.push(row.sales_qty || 0);
            stock.push(row.stock_qty || 0);

        });

        return {

            data: {

                labels: labels,

                datasets: [

                    {
                        name: "Sales",
                        values: sales
                    },

                    {
                        name: "Stock",
                        values: stock
                    }

                ]

            },

            type: "bar",

            height: 300

        };

    }

};

File: franchise_erp/franchise_erp/report/sales_vs_stock/sales_vs_stock.py
# # Copyright (c) 2026, Franchise Erp and contributors
# # For license information, please see license.txt

import frappe
from frappe.utils import cint


def execute(filters=None):
    filters = filters or {}

    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data)

    return columns, data, None, chart


def get_columns():
    return [
        {
            "label": "Style No.",
            "fieldname": "custom_barcode_code",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": "Department",
            "fieldname": "custom_department",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": "Month",
            "fieldname": "month",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": "Sales Qty",
            "fieldname": "sales_qty",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Stock Qty",
            "fieldname": "stock_qty",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Status",
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 120,
        },
    ]


def get_data(filters):
    # frappe.msgprint(str(filters))
    conditions = ["si.docstatus = 1"]
    params = {}

    # Date Filter
    if filters.get("from_date"):
        conditions.append("si.posting_date >= %(from_date)s")
        params["from_date"] = filters.get("from_date")

    if filters.get("to_date"):
        conditions.append("si.posting_date <= %(to_date)s")
        params["to_date"] = filters.get("to_date")

    # Company Filter
    if filters.get("company"):
        conditions.append("si.company = %(company)s")
        params["company"] = filters.get("company")

    # Month Filter ONLY when date filter not selected
    if (
        filters.get("month")
        and not filters.get("from_date")
        and not filters.get("to_date")
    ):
        conditions.append("MONTH(si.posting_date) = %(month)s")
        params["month"] = cint(filters.get("month"))

    where_clause = " AND ".join(conditions)

    data = frappe.db.sql(
        f"""
        SELECT

            i.image,
            i.custom_barcode_code,
            i.custom_departments AS custom_department,

            MONTH(si.posting_date) AS month,

            SUM(sii.qty) AS sales_qty,

            IFNULL(
                (
                    SELECT actual_qty
                    FROM `tabBin`
                    WHERE item_code = sii.item_code
                    LIMIT 1
                ),
                0
            ) AS stock_qty,

            CASE
                WHEN
                    IFNULL(
                        (
                            SELECT actual_qty
                            FROM `tabBin`
                            WHERE item_code = sii.item_code
                            LIMIT 1
                        ),
                        0
                    ) > SUM(sii.qty)
                THEN 'In Stock'

                ELSE 'Low Stock'

            END AS status

        FROM `tabSales Invoice Item` sii

        INNER JOIN `tabSales Invoice` si
            ON sii.parent = si.name

        LEFT JOIN `tabItem` i
            ON i.item_code = sii.item_code

        WHERE {where_clause}

        GROUP BY
            sii.item_code,
            MONTH(si.posting_date)

        ORDER BY
            sales_qty DESC
        """,
        params,
        as_dict=True,
    )

    for row in data:

        if row.get("image"):

            img = row.image

            if not img.startswith("/"):
                img = "/" + img

            row["image_url"] = img

        else:
            row["image_url"] = ""

        dept = row.get("custom_department") or ""

        row["custom_department"] = (
            dept.split("-")[-1].strip() if dept else ""
        )

    return data


def get_chart_data(data):

    return {
        "data": {
            "labels": [
                d.get("custom_barcode_code") or "No Style"
                for d in data
            ],
            "datasets": [
                {
                    "name": "Sales",
                    "values": [
                        float(d.get("sales_qty") or 0)
                        for d in data
                    ],
                },
                {
                    "name": "Stock",
                    "values": [
                        float(d.get("stock_qty") or 0)
                        for d in data
                    ],
                },
            ],
        },
        "type": "bar",
    }